File: numex-app/backend/app/schemas/debt.py
"""Debt schemas."""

from datetime import date
from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field, field_validator, model_validator


class DebtTypeEnum(str, Enum):
    CREDIT_CARD = "credit_card"
    MORTGAGE = "mortgage"
    CONSUMER_LOAN = "consumer_loan"
    CAR_LOAN = "car_loan"
    MICROLOAN = "microloan"


class PaymentTypeEnum(str, Enum):
    ANNUITY = "annuity"
    DIFFERENTIATED = "differentiated"
    MINIMUM_PERCENT = "minimum_percent"


class PrepaymentPolicyEnum(str, Enum):
    ALLOWED = "allowed"
    PROHIBITED = "prohibited"
    WITH_PENALTY = "with_penalty"


DEBT_TYPE_DEFAULTS: dict[DebtTypeEnum, dict] = {
    DebtTypeEnum.CREDIT_CARD: {
        "payment_type": PaymentTypeEnum.MINIMUM_PERCENT,
        "prepayment_policy": PrepaymentPolicyEnum.ALLOWED,
        "term_months": None,
        "min_payment_pct": 5.0,
    },
    DebtTypeEnum.MORTGAGE: {
        "payment_type": PaymentTypeEnum.ANNUITY,
        "prepayment_policy": PrepaymentPolicyEnum.WITH_PENALTY,
        "term_months": 240,
        "min_payment_pct": 0,
    },
    DebtTypeEnum.CONSUMER_LOAN: {
        "payment_type": PaymentTypeEnum.ANNUITY,
        "prepayment_policy": PrepaymentPolicyEnum.ALLOWED,
        "term_months": 36,
        "min_payment_pct": 0,
    },
    DebtTypeEnum.CAR_LOAN: {
        "payment_type": PaymentTypeEnum.ANNUITY,
        "prepayment_policy": PrepaymentPolicyEnum.WITH_PENALTY,
        "term_months": 60,
        "min_payment_pct": 0,
    },
    DebtTypeEnum.MICROLOAN: {
        "payment_type": PaymentTypeEnum.ANNUITY,
        "prepayment_policy": PrepaymentPolicyEnum.ALLOWED,
        "term_months": 6,
        "min_payment_pct": 0,
    },
}


class DebtCreate(BaseModel):
    name: str
    debt_type: DebtTypeEnum

    principal: float
    current_balance: float | None = Field(default=None, description="Defaults to principal if omitted")
    interest_rate_annual: float

    payment_type: PaymentTypeEnum | None = None
    min_payment_pct: float | None = None
    fixed_payment: float | None = None

    prepayment_policy: PrepaymentPolicyEnum | None = None
    prepayment_penalty_pct: float | None = None

    late_fee_rate: float = 0
    start_date: date
    term_months: int | None = Field(default=None, description="Not used for credit_card")

    credit_limit: float | None = None
    grace_period_days: int | None = None

    @field_validator("term_months", mode="before")
    @classmethod
    def term_months_allow_none(cls, v: object) -> int | None:
        if v is None:
            return None
        return v

    @model_validator(mode="before")
    @classmethod
    def inject_missing_optionals(cls, data: dict) -> dict:
        if isinstance(data, dict):
            for key in ("current_balance", "term_months"):
                if key not in data:
                    data = {**data, key: None}
        return data

    @model_validator(mode="after")
    def apply_defaults_and_validate(self) -> "DebtCreate":
        defaults = DEBT_TYPE_DEFAULTS.get(self.debt_type, {})

        updates: dict = {}
        if self.current_balance is None:
            updates["current_balance"] = self.principal
        if self.payment_type is None:
            updates["payment_type"] = defaults.get("payment_type", PaymentTypeEnum.ANNUITY)
        if self.prepayment_policy is None:
            updates["prepayment_policy"] = defaults.get("prepayment_policy", PrepaymentPolicyEnum.ALLOWED)
        if self.term_months is None:
            updates["term_months"] = defaults.get("term_months")
        if self.min_payment_pct is None:
            updates["min_payment_pct"] = defaults.get("min_payment_pct", 0)
        if self.debt_type == DebtTypeEnum.CREDIT_CARD and self.credit_limit is None:
            updates["credit_limit"] = self.principal
        if self.prepayment_policy == PrepaymentPolicyEnum.WITH_PENALTY and self.prepayment_penalty_pct is None:
            updates["prepayment_penalty_pct"] = 1.0

        if not updates:
            return self
        return self.model_copy(update=updates)


class DebtUpdate(BaseModel):
    name: str | None = None
    current_balance: float | None = None
    interest_rate_annual: float | None = None
    min_payment_pct: float | None = None
    fixed_payment: float | None = None
    prepayment_policy: PrepaymentPolicyEnum | None = None
    prepayment_penalty_pct: float | None = None
    term_months: int | None = None
    credit_limit: float | None = None
    grace_period_days: int | None = None


class DebtResponse(BaseModel):
    id: str
    name: str
    debt_type: DebtTypeEnum

    principal: float
    current_balance: float
    interest_rate_annual: float

    payment_type: PaymentTypeEnum
    min_payment_pct: float
    fixed_payment: float | None

    prepayment_policy: PrepaymentPolicyEnum
    prepayment_penalty_pct: float | None

    late_fee_rate: float | None
    start_date: date
    term_months: int | None

    credit_limit: float | None
    grace_period_days: int | None

    model_config = {"from_attributes": True}


DEBT_TYPE_LABELS: dict[DebtTypeEnum, str] = {
    DebtTypeEnum.CREDIT_CARD: "Кредитная карта",
    DebtTypeEnum.MORTGAGE: "Ипотека",
    DebtTypeEnum.CONSUMER_LOAN: "Потребительский кредит",
    DebtTypeEnum.CAR_LOAN: "Автокредит",
    DebtTypeEnum.MICROLOAN: "Микрозайм",
}

File: numex-app/frontend/src/types/debt.ts
export type DebtType = 'credit_card' | 'mortgage' | 'consumer_loan' | 'car_loan' | 'microloan'
export type PaymentType = 'annuity' | 'differentiated' | 'minimum_percent'
export type PrepaymentPolicy = 'allowed' | 'prohibited' | 'with_penalty'

export interface Debt {
  id: string
  name: string
  debt_type: DebtType

  principal: number
  current_balance: number
  interest_rate_annual: number

  payment_type: PaymentType
  min_payment_pct: number
  fixed_payment: number | null

  prepayment_policy: PrepaymentPolicy
  prepayment_penalty_pct: number | null

  late_fee_rate: number | null
  start_date: string
  term_months: number | null

  credit_limit: number | null
  grace_period_days: number | null
}

export interface DebtCreate {
  name: string
  debt_type: DebtType

  principal: number
  current_balance?: number
  interest_rate_annual: number

  payment_type?: PaymentType
  min_payment_pct?: number
  fixed_payment?: number

  prepayment_policy?: PrepaymentPolicy
  prepayment_penalty_pct?: number

  late_fee_rate?: number
  start_date: string
  term_months?: number | null

  credit_limit?: number
  grace_period_days?: number
}

export const DEBT_TYPE_LABELS: Record<DebtType, string> = {
  credit_card: 'Кредитная карта',
  mortgage: 'Ипотека',
  consumer_loan: 'Потребительский кредит',
  car_loan: 'Автокредит',
  microloan: 'Микрозайм',
}

export const PAYMENT_TYPE_LABELS: Record<PaymentType, string> = {
  annuity: 'Аннуитетный',
  differentiated: 'Дифференцированный',
  minimum_percent: '% от остатка',
}

export const PREPAYMENT_LABELS: Record<PrepaymentPolicy, string> = {
  allowed: 'Разрешено',
  prohibited: 'Запрещено',
  with_penalty: 'Со штрафом',
}

export const DEBT_TYPE_DEFAULTS: Record<DebtType, Partial<DebtCreate>> = {
  credit_card: {
    payment_type: 'minimum_percent',
    prepayment_policy: 'allowed',
    min_payment_pct: 5,
    term_months: undefined,
  },
  mortgage: {
    payment_type: 'annuity',
    prepayment_policy: 'with_penalty',
    min_payment_pct: 0,
    term_months: 240,
  },
  consumer_loan: {
    payment_type: 'annuity',
    prepayment_policy: 'allowed',
    min_payment_pct: 0,
    term_months: 36,
  },
  car_loan: {
    payment_type: 'annuity',
    prepayment_policy: 'with_penalty',
    min_payment_pct: 0,
    term_months: 60,
  },
  microloan: {
    payment_type: 'annuity',
    prepayment_policy: 'allowed',
    min_payment_pct: 0,
    term_months: 6,
  },
}

export interface DebtCardData {
  id: string
  name: string
  totalAmount: number
  remainingAmount: number
  monthlyPayment: number
  rate: number
  nextPaymentDate: string
  minPayment: number
  type: DebtType
}

export interface OptimizationPlan {
  id?: string
  total_cost: number
  savings_vs_minimum: number | null
  baseline_cost?: number
  payments_matrix: Record<string, number[]>
  balances_matrix?: Record<string, number[]>
  status?: string
  solve_time?: number
  horizon_months?: number
  created_at?: string
}

export interface OptimizationRequest {
  monthly_budget?: number
  budget_by_month?: number[]
  horizon_months?: number
}

export function mapDebtToCardData(debt: Debt): DebtCardData {
  let monthlyPayment: number
  if (debt.fixed_payment) {
    monthlyPayment = debt.fixed_payment
  } else if (debt.payment_type === 'minimum_percent') {
    monthlyPayment = Math.ceil(debt.current_balance * (debt.min_payment_pct / 100))
  } else {
    monthlyPayment = calculateAnnuityPayment(
      debt.current_balance,
      debt.interest_rate_annual,
      debt.term_months || 12
    )
  }

  const nextPaymentDate = new Date()
  nextPaymentDate.setDate(nextPaymentDate.getDate() + 30)

  return {
    id: debt.id,
    name: debt.name,
    totalAmount: debt.principal,
    remainingAmount: debt.current_balance,
    monthlyPayment,
    rate: debt.interest_rate_annual,
    nextPaymentDate: nextPaymentDate.toISOString().split('T')[0],
    minPayment: monthlyPayment,
    type: debt.debt_type,
  }
}

function calculateAnnuityPayment(principal: number, annualRate: number, termMonths: number): number {
  const monthlyRate = Math.pow(1 + annualRate / 100, 1 / 12) - 1
  if (monthlyRate === 0) {
    return Math.ceil(principal / termMonths)
  }
  return Math.ceil(
    principal * (monthlyRate * Math.pow(1 + monthlyRate, termMonths)) /
    (Math.pow(1 + monthlyRate, termMonths) - 1)
  )
}
